_posts/2021-12-29-readable-tests-by-example.md

```diff
@@ -0,0 +1,552 @@
+to the domain language that obscure the presence of natural expressions that we use in conversations with business
+stakeholders, for instance “event publisher” or “mock server”.
+
+Another disadvantage of this code is that it is not easily adaptable to further development, e.g. in the event of a
+change in business requirements when it is necessary to modify or add another test.
+
+The conscious reader might notice that the example of our imperfect test is maybe too exaggerated and that each section
+of the ‘given’/‘when’/’then’ blocks, etc. could be extracted by the use of a separate private method. Certainly, this
+procedure may result in some improvement of the code quality, but nevertheless such an approach still has many
+drawbacks:
+
+* the test class still contains code related to the technical implementation;
+* if another test class uses a similar subset of functionalities, then sooner or later, there will be a need to copy
+  such a method;
+* what if I would like to change, for example, the library for mocking calls to another type of library?
+
+Let’s take a closer look at a unit test this time. It covers a narrower range of requirements because, e.g., it does not
+check whether the client has been sent a free music track. Try to find similar defects in it as in the acceptance test.
+
+```groovy
+final Money EUR_40 = Money.of("40.00", "EUR")
+final ClientId CLIENT_ID = new ClientId("1")
+final Vinyl VINYL_1 = new Vinyl(new VinylId("1"), EUR_40)
+final Quantity ONE = new Quantity(1)
+final OrderId ORDER_ID = new OrderId("1")
+final OrderDataSnapshot UNPAID_ORDER_EUR_40 = orderFactory.create(ORDER_ID, CLIENT_ID, Maps.of(VINYL_1, ONE), true)
+    .toSnapshot()
+final ClientReputation VIP = ClientReputation.vip(CLIENT_ID)
+final PayOrderCommand PAY_FOR_ORDER_EUR_40 = new PayOrderCommand(ORDER_ID, EUR_40)
+
+def "shouldn’t charge for delivery when the client has a VIP status"() {
+    given:
+        orderRepository.findBy(ORDER_ID) >> Optional.of(UNPAID_ORDER_EUR_40)
+
+    and:
+        clientReputationProvider.get(CLIENT_ID) >> VIP
+
+    when:
+        def result = paymentHandler.handle(PAY_FOR_ORDER_EUR_40)
+
+    then:
+        result.isSuccess()
+
+    and:
+        1 * domainEventPublisher.publish({ OrderPaid event ->
+            assert event.orderId() == ORDER_ID
+            assert event.amount() == EUR_40
+            assert event.delivery().cost() == Money.ZERO
+            assert event.when() == CURRENT_DATE
+        })
+}
+```
+
+In this case, it may seem that test is much better, because it is simpler and easier to read, but with a more in-depth
+analysis it turns out that it still does not meet the expected requirements from the “How can we test such
+code?” section, because:
+
+* it uses concepts such as repository, provider, event publisher, handler, which were not mentioned in the written business
+  scenarios - these are technical implementation details;
```

Comment:
https://github.com/allegro/blog/blob/main/CONTRIBUTING.md#dont-confuse-hyphens-and-dashes
tutaj użyłabym myślnika (em dash) zamiast dywizu, zgodnie z naszym "contributing guide". Dodałam dodatkowe komentarze, gdzie jeszcze użyłabym myślnika 👇

Reply:
Done